File: TsinghuaNet.UWP.Background/LoginTask.vb
Imports TsinghuaNet.Helper
Imports TsinghuaNet.UWP.Helper
Imports Windows.ApplicationModel.Background

Public NotInheritable Class LoginTask
    Implements IBackgroundTask

    Public Async Sub Run(taskInstance As IBackgroundTaskInstance) Implements IBackgroundTask.Run
        Dim deferral = taskInstance.GetDeferral()
        Try
            Dim status As New InternetStatus
            Await status.RefreshAsync()
            Dim un = SettingsHelper.StoredUsername
            Dim pw = CredentialHelper.GetCredential(un)
            Dim helper = ConnectHelper.GetHelper(Await status.SuggestAsync(), un, pw)
            If helper IsNot Nothing Then
                Using helper
                    Await helper.LoginAsync()
                    Dim user As FluxUser = Await helper.GetFluxAsync()
                    NotificationHelper.UpdateTile(user)
                    NotificationHelper.SendToast(user)
                    If SettingsHelper.FluxLimit IsNot Nothing Then
                        NotificationHelper.SendWarningToast(user, SettingsHelper.FluxLimit)
                    End If
                End Using
            End If
        Finally
            deferral.Complete()
        End Try
    End Sub
End Class

File: TsinghuaNet.Uno/TsinghuaNet.Uno.UWP.Background/LoginTask.cs
using TsinghuaNet.Models;
using TsinghuaNet.Uno.Helpers;
using Windows.ApplicationModel.Background;

namespace TsinghuaNet.Uno.UWP.Background
{
    public sealed class LoginTask : IBackgroundTask
    {
        public async void Run(IBackgroundTaskInstance taskInstance)
        {
            var deferral = taskInstance.GetDeferral();
            try
            {
                InternetStatus status = new InternetStatus();
                NetUnoSettings settings = new NetUnoSettings();
                settings.LoadSettings();
                await status.RefreshAsync();
                var un = settings.StoredUsername;
                var pw = CredentialHelper.GetCredential(un);
                var helper = ConnectHelper.GetHelper(await status.SuggestAsync(), un, pw);
                if (helper != null)
                {
                    using (helper)
                    {
                        await helper.LoginAsync();
                        FluxUser user = await helper.GetFluxAsync();
                        NotificationHelper.UpdateTile(user);
                        NotificationHelper.SendToast(user);
                        if (settings.EnableFluxLimit)
                            NotificationHelper.SendWarningToast(user, settings.FluxLimit);
                    }
                }
            }
            finally
            {
                deferral.Complete();
            }
        }
    }
}
